BugFix 11064: st.dataframe({}) shouldn't show errors

sfc-gh-nbellante opened the pull request:
## Describe your changes
When an empty dataframe is constructed, the user sees an error message that says it should never happen. This PR makes an exception for the case where `numRows` is 0.

## GitHub Issue Link (if applicable)
closes #11064 

## Testing Plan

added e2e tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
Add handling for empty dataframes in useDataLoader hook

- Implemented a check in the getCellContent function to return an empty cell when the dataframe has no rows, improving the handling of empty dataframes.
- Imported the getEmptyCell utility to facilitate this functionality.

sfc-gh-nbellante committed:
adding test

sfc-gh-bnisco commented on e2e_playwright/st_dataframe_input_data_test.py:
```diff
@@ -42,3 +46,33 @@ def test_dataframe_input_format_rendering(
+        # Hover over the first row
+        dataframe_element.hover(position={"x": first_cell_x, "y": first_row_y})
+
+        # Wait and verify no error tooltips appear
+        wait_until(
+            app,
+            lambda: not app.get_by_text("This error should never happen").is_visible(),
+        )
```
**question:** Is this check sufficient enough to ensure this message never shows up? Checking for the non-existence of things can be difficult, especially if it relies on timing. Perhaps one way to verify is if you removed the fix in `useDataLoader`, does this test fail?

sfc-gh-nbellante replied:
good call! turns out it wasn't working as it was checking before the tooltip could pop up. i added a sleep and increased it until it would 100% of the time fail, and now it seems to work as expected

lukasmasuch commented on frontend/lib/src/components/widgets/DataFrame/hooks/useDataLoader.ts:
```diff
@@ -48,6 +49,12 @@ function useDataLoader(
   const getCellContent = React.useCallback(
     ([col, row]: readonly [number, number]): GridCell => {
+      // Special case for empty dataframes
+      if (numRows === 0) {
+        // Return an empty cell for empty dataframes
+        return getEmptyCell()
+      }
```
Solving this using the `ignoredRowIndices` parameter of `useTooltips` would be a bit more consistent. 
The `getCellContent` is only supposed to be called for actual data cells; everything else is expected to return an error cell since it's an unexpected situation. The `empty` row of a dataframe with 0 rows isn't part of the actual data, it's just a visual gimmick we add that shouldn't cause the tooltip implementation to trigger a getCellContent call. It is similar to the trailing row that is added in the data editor, which is ignored via the `ignoredRowIndices` parameter. In this case, we can just do something like: `isEmptyTable ? [0] : isDynamicAndEditable ? [numRows] : []` when calling `useTooltips`.

sfc-gh-nbellante committed:
changing the fix to be how lukas suggested

sfc-gh-nbellante committed:
lint

sfc-gh-nbellante committed:
fix test

sfc-gh-nbellante committed:
still use wait until, but keep the sleep

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -546,11 +546,12 @@ function DataFrame({
     // The index are 0-based -> therefore, numRows will point to the trailing row
     // (which is not part of the actual data).
-    isDynamicAndEditable ? [numRows] : []
+    isEmptyTable ? [0] : isDynamicAndEditable ? [numRows] : []
```
**suggestion (non-blocking):** Optimization: creating arrays inline like this generates a new instance on every render. Even if the contents are identical, the changing reference will cause hooks in `useTooltips` to re-calculate unnecessarily.

We could wrap the array creation in `useMemo` to ensure a stable reference when the underlying values haven't changed. This might also be a good opportunity to potentially simplify the nested ternary logic within the `useMemo` callback for better readability.

sfc-gh-nbellante committed:
Refactor tooltip handling in DataFrame component

- Moved the calculation of ignored row indices into a memoized function for better readability and performance.
- Updated the useTooltips hook to utilize the new ignoredRowIndices logic, improving tooltip behavior for empty and dynamically editable dataframes.
